File: translation_output/w_test/wizardcoder/codenet/java/python/p03365.py
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Translate the following java code to python and enclose your solution inside ```python```.
A sample test case is provided below:

Test input:
4

Expected output:
16


```
import java.io.OutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.PrintWriter;
import java.util.InputMismatchException;
import java.io.IOException;
import java.io.InputStream;

/**
 * Built using CHelper plug-in
 * Actual solution is at the top
 */
public class p03365 {
    public static void main(String[] args) {
        InputStream inputStream = System.in;
        OutputStream outputStream = System.out;
        MyScan in = new MyScan(inputStream);
        PrintWriter out = new PrintWriter(outputStream);
        TaskC solver = new TaskC();
        solver.solve(1, in, out);
        out.close();
    }

    static class TaskC {
        public void solve(int testNumber, MyScan in, PrintWriter out) {
            int i = in.nextInt();

            long[] md = new long[i];

            int rm = i - 2;
            int lx = i - 2;
            while (rm >= 0) {
                md[lx + 1] = Util.c(rm, lx);
                rm -= 2;
                lx--;
            }
            long res = 0;
            long prev = 0;
            for (int m = 1; m < i; m++) {
                long cur = Util.fact(m) * Util.fact(i - m - 1) % Util._m;
                cur = cur * (md[m]) % Util._m;


                res += (cur - prev + Util._m) * m % Util._m;
                prev = cur;

            }


            out.println(res % Util._m);

        }

    }

    static class MyScan {
        private final InputStream in;
        private byte[] inbuf = new byte[1024];
        public int lenbuf = 0;
        public int ptrbuf = 0;

        public MyScan(InputStream in) {
            this.in = in;
        }

        private int readByte() {
            if (lenbuf == -1) throw new InputMismatchException();
            if (ptrbuf >= lenbuf) {
                ptrbuf = 0;
                try {
                    lenbuf = in.read(inbuf);
                } catch (IOException e) {
                    throw new InputMismatchException();
                }
                if (lenbuf <= 0) return -1;
            }
            return inbuf[ptrbuf++];
        }

        public int nextInt() {
            int num = 0, b;
            boolean minus = false;
            while ((b = readByte())!= -1 &&!((b >= '0' && b <= '9') || b == '-')) ;
            if (b == '-') {
                minus = true;
                b = readByte();
            }

            while (true) {
                if (b >= '0' && b <= '9') {
                    num = num * 10 + (b - '0');
                } else {
                    return minus? -num : num;
                }
                b = readByte();
            }
        }

    }

    static class Util {
        public static final long M07 = 1000_000_007;
        public static long _m = M07;
        private static long[] fact;
        private static long[] inv;
        private static long[] ifact;

        public static void l() {
            fact = new long[1000001];
            inv = new long[1000001];
            ifact = new long[1000001];
            for (int s = 0; s < 2; s++) {
                fact[s] = 1;
                ifact[s] = 1;
                inv[s] = 1;
            }
            inv[0] = 0;
            for (int i = 2; i < inv.length; i++) {
                inv[i] = (_m - (_m / i) * inv[(int) (_m % i)] % _m) % _m;
            }
            for (int i = 2; i < fact.length; i++) {
                fact[i] = i * fact[i - 1] % _m;
                ifact[i] = inv[i] * ifact[i - 1] % _m;
            }
        }

        public static long c(int a, int b) {
            if (fact == null) l();
            return ifact[a] * ifact[b - a] % _m * fact[b] % _m;
        }

        public static long fact(int a) {
            if (fact == null) l();
            return fact[a];
        }

    }
}


```

### Response:
```python
import java.io.OutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.PrintWriter;
import java.util.InputMismatchException;
import java.io.IOException;
import java.io.InputStream;

class TaskC:
    def solve(self, testNumber, in, out):
        i = in.nextInt()

        md = [0] * i

        rm = i - 2
        lx = i - 2
        while rm >= 0:
            md[lx + 1] = Util.c(rm, lx)
            rm -= 2
            lx -= 1

        res = 0
        prev = 0
        for m in range(1, i):
            cur = Util.fact(m) * Util.fact(i - m - 1) % Util._m
            cur = cur * (md[m]) % Util._m

            res += (cur - prev + Util._m) * m % Util._m
            prev = cur

        out.println(res % Util._m)

class Util:
    M07 = 1000_000_007
    _m = M07
    fact = None
    inv = None
    ifact = None

    @staticmethod
    def l():
        Util.fact = [1] * 1000001
        Util.inv = [1] * 1000001
        Util.ifact = [1] * 1000001
        for s in range(2):
            Util.fact[s] = 1
            Util.ifact[s] = 1
            Util.inv[s] = 1
        Util.inv[0] = 0
        for i in range(2, Util.inv.length):
            Util.inv[i] = (Util._m - (Util._m // i) * Util.inv[Util._m % i] % Util._m) % Util._m
        for i in range(2, Util.fact.length):
            Util.fact[i] = i * Util.fact[i - 1] % Util._m
            Util.ifact[i] = Util.inv[i] * Util.ifact[i - 1] % Util._m

    @staticmethod
    def c(a, b):
        if Util.fact is None:
            Util.l()
        return Util.ifact[a] * Util.ifact[b - a] % Util._m * Util.fact[b] % Util._m

    @staticmethod
    def fact(a):
        if Util.fact is None:
            Util.l()
        return Util.fact[a]

class MyScan:
    def __init__(self, in):
        self.in = in
        self.inbuf = bytearray(1024)
        self.lenbuf = 0
        self.ptrbuf = 0

    def readByte(self):
        if self.lenbuf == -1:
            raise InputMismatchException()
        if self.ptrbuf >= self.lenbuf:
            self.ptrbuf = 0
            try:
                self.lenbuf = self.in.read(self.inbuf)
            except IOException as e:
                raise InputMismatchException()
            if self.lenbuf <= 0:
                return -1
        return self.inbuf[self.ptrbuf]

    def nextInt(self):
        num = 0
        b = self.readByte()
        minus = False
        while b!= -1 and not ((b >= '0' and b <= '9') or b == '-'):
            b = self.readByte()
        if b == '-':
            minus = True
            b = self.readByte()
        while b!= -1 and (b >= '0' and b <= '9'):
            num = num * 10 + (b - '0')
            b = self.readByte()
        return -num if minus else num

if __name__ == '__main__':
    inputStream = System.in
    outputStream = System.out
    in = MyScan(inputStream)
    out = PrintWriter(outputStream)
    solver = TaskC()
    solver.solve(1, in, out)
    out.close()
```<|endoftext|>

File: translation_output/w_test/wizardcoder/codenet/python/java/p03365.java
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Translate the following python code to java and enclose your solution inside ```java```.
A sample test case is provided below:

Test input:
4

Expected output:
16


```
#!/usr/bin/python
# -*- coding: utf-8 -*-
import sys,collections,itertools,re,math,fractions,decimal,random,array,bisect,heapq

# decimal.getcontext().prec = 50
# sys.setrecursionlimit(10000)
MOD = 10**9 + 7

def solve(f):
    n = f.read_int()

    fact = [1] * (n+1)
    ifact = [1] * (n+1)
    for i in range(1, n+1):
        fact[i] = (fact[i-1] * i) % MOD
        ifact[i] = pow(fact[i], MOD-2, MOD)

    def nCk(n, k):
        if k < 0 or k > n: return 0
        return (fact[n] * ifact[n-k] * ifact[k]) % MOD

    ans = 0
    c0 = 0
    for i in xrange(n/2,n):
        c1 = (nCk(i-1, n-1-i)*fact[i]*fact[n-i-1] - c0 + MOD) % MOD
        ans = (ans + c1*i) % MOD
        c0 = (c0 + c1) % MOD

    return ans


class Reader(object):
    def __init__(self, filename=None):
        self.file = open(filename) if filename is not None else None
        self.case = 1

    def __readline(self):
        return self.file.next().strip() if self.file else raw_input()

    def next_case(self):
        self.file.next()
        self.case += 1

    def read_int(self): return int(self.__readline())
    def read_float(self): return float(self.__readline())
    def read_long(self): return long(self.__readline())
    def read_decimal(self): return decimal.Decimal(self.__readline())
    def read_str(self): return self.__readline()

    def read_int_list(self): return map(int, self.__readline().split())
    def read_float_list(self): return map(float, self.__readline().split())
    def read_long_list(self): return map(long, self.__readline().split())
    def read_decimal_list(self): return map(decimal.Decimal, self.__readline().split())
    def read_str_list(self): return self.__readline().split()

if __name__ == '__main__':
    filename = sys.argv[1] if len(sys.argv) > 1 else None
    f = Reader(filename)
    if f.file:
        while True:
            print "Case #%d\n"%f.case, solve(f)
            try:
                f.next_case()
            except StopIteration:
                break
    else:
        print solve(f)

```

### Response:
```java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int t = Integer.parseInt(br.readLine());
        for (int i = 1; i <= t; i++) {
            int n = Integer.parseInt(br.readLine());

            long[] fact = new long[n+1];
            fact[0] = 1;
            for (int j = 1; j <= n; j++) {
                fact[j] = (fact[j-1] * j) % 1000000007;
            }

            long[] ifact = new long[n+1];
            ifact[n] = pow(fact[n], 1000000006, 1000000007);
            for (int j = n-1; j >= 0; j--) {
                ifact[j] = (ifact[j+1] * (j+1)) % 1000000007;
            }

            long ans = 0;
            long c0 = 0;
            for (int j = n/2; j < n; j++) {
                long c1 = (nCk(j-1, n-1-j)*fact[j]*fact[n-j-1] - c0 + 1000000007) % 1000000007;
                ans = (ans + c1*j) % 1000000007;
                c0 = (c0 + c1) % 1000000007;
            }

            bw.write("Case #" + i + ": " + ans + "\n");
        }

        bw.close();
    }

    public static long pow(long a, long b, long mod) {
        long res = 1;
        while (b > 0) {
            if (b % 2 == 1) res = (res * a) % mod;
            a = (a * a) % mod;
            b /= 2;
        }
        return res;
    }

    public static long nCk(long n, long k) {
        if (k < 0 || k > n) return 0;
        return (fact[n] * ifact[n-k] * ifact[k]) % 1000000007;
    }

    public static long[] fact = new long[1000001];
    public static long[] ifact = new long[1000001];
}
```<|endoftext|>
